File: MY_TWIN_PRO/MY-TWIN-PRO/src/services/SubscriptionService.ts
import { commercePlugin, PlanTier, PurchaseResult } from './CommercePlugin';
import { EventBus } from '../core/EventBus';
import { StateBus } from '../core/StateBus';

const TIER_CAPABILITIES: Record<PlanTier, string[]> = {
  free: ['chat', 'weather', 'search', 'translate', 'summarize'],
  plus: ['chat', 'study', 'content', 'dreams', 'proactive'],
  premium: ['chat', 'study', 'code', 'business', 'coach', 'content', 'dreams', 'proactive', 'deep_search', 'shadow_mode'],
  pro: ['chat', 'study', 'code', 'business', 'coach', 'content', 'dreams', 'smart_home', 'proactive', 'deep_search', 'shadow_mode'],
  yearly: ['all'],
};

export class SubscriptionService {
  private currentTier: PlanTier = 'free';
  private capabilities: string[] = [];
  private isInitialized = false;

  async initialize(userId: string): Promise<void> {
    if (this.isInitialized) return;
    await commercePlugin.initialize();
    const subscription = await commercePlugin.getCurrentSubscription(userId);
    if (subscription?.isActive) {
      this.currentTier = subscription.tier;
    } else {
      this.currentTier = 'free';
    }
    this.capabilities = TIER_CAPABILITIES[this.currentTier] || TIER_CAPABILITIES.free;
    this.isInitialized = true;
    EventBus.emit('SUBSCRIPTION_INITIALIZED', { tier: this.currentTier, capabilities: this.capabilities });
  }

  async purchase(planId: PlanTier, userId: string): Promise<PurchaseResult> {
    const result = await commercePlugin.purchase(planId, userId);
    if (result.success) {
      this.currentTier = planId;
      this.capabilities = TIER_CAPABILITIES[planId] || TIER_CAPABILITIES.free;
      EventBus.emit('SUBSCRIPTION_UPDATED', { tier: this.currentTier, capabilities: this.capabilities });
    }
    return result;
  }

  async restore(userId: string): Promise<PurchaseResult> {
    const result = await commercePlugin.restorePurchases(userId);
    if (result.success && result.tier) {
      this.currentTier = result.tier;
      this.capabilities = TIER_CAPABILITIES[result.tier] || TIER_CAPABILITIES.free;
      EventBus.emit('SUBSCRIPTION_RESTORED', { tier: this.currentTier, capabilities: this.capabilities });
    }
    return result;
  }

  /** ✅ ترقية مؤقتة لمدة محدودة (للمفاجآت) */
  async upgradeForDuration(userId: string, tier: PlanTier, days: number): Promise<boolean> {
    try {
      const { apiPost } = require('../../lib/httpClient');
      const res = await apiPost('/api/billing/upgrade-temporary', {
        user_id: userId,
        tier: tier,
        duration_days: days,
      });
      if (res?.success) {
        this.currentTier = tier;
        this.capabilities = TIER_CAPABILITIES[tier] || TIER_CAPABILITIES.free;
        EventBus.emit('SUBSCRIPTION_UPDATED', { tier: this.currentTier, capabilities: this.capabilities });
        return true;
      }
      return false;
    } catch (e) {
      console.warn('[Subscription] Temporary upgrade failed:', e);
      return false;
    }
  }

  canUseCapability(capability: string): boolean {
    if (this.capabilities.includes('all')) return true;
    return this.capabilities.includes(capability);
  }

  getCurrentTier(): PlanTier { return this.currentTier; }
  getCapabilities(): string[] { return [...this.capabilities]; }
  isPremium(): boolean { return this.currentTier !== 'free'; }

  async cancel(userId: string): Promise<boolean> {
    const success = await commercePlugin.cancel(userId);
    if (success) {
      this.currentTier = 'free';
      this.capabilities = TIER_CAPABILITIES.free;
      EventBus.emit('SUBSCRIPTION_CANCELLED', {});
    }
    return success;
  }
}

export const subscriptionService = new SubscriptionService();

File: MY_TWIN_PRO/MY-TWIN-PRO/backend/app/domain/billing/subscription_service.py
"""
Subscription Service v4.0 – مع استعادة الطاقة عند الترقية
"""
import logging
from typing import Dict, Any, Optional
from datetime import datetime, timezone, timedelta
from app.infrastructure.database.supabase_client import get_db

logger = logging.getLogger("subscription_service")

SUBSCRIPTION_PLANS = {
    "free":    {"name": "Free",    "price": 0,     "messages": 15,   "features": ["chat", "weather", "search", "translate", "summarize"]},
    "plus":    {"name": "Plus",    "price": 5.99,  "messages": 50,   "features": ["chat","study","content","dreams","proactive"]},
    "premium": {"name": "Premium", "price": 14.99, "messages": 150,  "features": ["chat","study","code","business","coach","content","dreams","proactive","deep_search","shadow_mode"]},
    "pro":     {"name": "Pro",     "price": 110,   "messages": 500,  "billing_period":"6_months","features":["chat","study","code","business","coach","content","dreams","smart_home","proactive","deep_search","shadow_mode"]},
    "yearly":  {"name": "Yearly",  "price": 199,   "messages": 9999, "billing_period":"yearly","features":["all"]},
}

async def get_user_subscription(user_id: str) -> Dict[str, Any]:
    db = get_db()
    try:
        r = db.table("profiles").select("tier,subscription_expires,subscription_id").eq("id", user_id).single().execute()
        if r.data:
            tier = r.data.get("tier", "free")
            plan = SUBSCRIPTION_PLANS.get(tier, SUBSCRIPTION_PLANS["free"]).copy()
            is_active = await check_subscription_active(user_id)
            return {"tier": tier, "plan": plan, "expires_at": r.data.get("subscription_expires"), "is_active": is_active, "subscription_id": r.data.get("subscription_id")}
    except Exception as e:
        logger.error(f"فشل جلب الاشتراك: {e}")
    return {"tier": "free", "plan": SUBSCRIPTION_PLANS["free"], "is_active": True}

async def check_subscription_active(user_id: str) -> bool:
    sub = await get_user_subscription(user_id)
    if sub["tier"] == "free": return True
    if sub.get("expires_at"):
        try:
            expires = datetime.fromisoformat(sub["expires_at"].replace("Z", "+00:00"))
            return expires > datetime.now(timezone.utc)
        except: return True
    return True

async def upgrade_subscription(user_id: str, tier: str, duration_days: int = 30) -> bool:
    db = get_db()
    try:
        expires = (datetime.now(timezone.utc) + timedelta(days=duration_days)).isoformat()
        db.table("profiles").update({
            "tier": tier, "subscription_expires": expires, "subscription_id": f"sub_{user_id[:8]}_{tier}"
        }).eq("id", user_id).execute()
        try:
            from app.engine.energy.twin_energy_engine import twin_energy_engine
            await twin_energy_engine.restore_energy(user_id, amount=1.0, source="upgrade")
            logger.info(f"⚡ طاقة {user_id} استعيدت بالكامل بعد الترقية إلى {tier}")
        except Exception as e:
            logger.warning(f"فشل استعادة الطاقة عند الترقية: {e}")
        try:
            from app.infrastructure.cache.cache_service import set as cache_set
            cache_set(f"tier:{user_id}", tier, 600)
        except Exception:
            pass
        try:
            from app.memory.emotional.emotional_memory import store_emotional_memory
            await store_emotional_memory(
                user_id=user_id, expressed_text=f"تمت ترقية الاشتراك إلى {tier}",
                detected_emotion={"primary": "joy", "intensity": 0.9, "valence": 0.8}, trigger="subscription_upgraded"
            )
        except: pass
        logger.info(f"✅ تمت ترقية المستخدم {user_id} إلى {tier}")
        return True
    except Exception as e:
        logger.error(f"فشل ترقية الاشتراك: {e}")
        return False


async def get_feature_access(user_id: str, feature: str) -> bool:
    sub = await get_user_subscription(user_id)
    if not sub.get("is_active", False): return False
    features = sub.get("plan", {}).get("features", [])
    return "all" in features or feature in features

logger.info("✅ Subscription Service v4.0 initialized with energy restore")
